### ISSUE TITLE
DRA: DRAResourceClaimDeviceStatus API cleanup

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#128831]
### What would you like to be added?

https://github.com/kubernetes/kubernetes/pull/128240 added claim.status.devices. Some cleanup of the API would be useful.

- Fix inconsistency:

```console
$ diff -c pkg/apis/resource/types.go <(sed -e 's;`json:",inline.*;// inline;' -e 's;metav1.TypeMeta // inline;metav1.TypeMeta;' -e 's; `json.*;;' -e 's/v1\./core./' -e's/metacore./metav1./' -e 's/package v1beta1/package resource/' staging/src/k8s.io/api/resource/v1beta1/types.go | grep -v -e '+genclient' -e '+k8s:prerelease-lifecycle-gen' )
*** pkg/apis/resource/types.go	2024-11-18 10:12:21.388362592 +0100
--- /dev/fd/63	2024-11-18 10:17:52.573838086 +0100
...
***************
*** 1020,1027 ****
  	// If the device has been configured according to the class and claim
  	// config references, the `Ready` condition should be True.
  	//
- 	// Must not contain more than 8 entries.
- 	//
  	// +optional
  	// +listType=map
  	// +listMapKey=type
--- 1020,1025 ----
***************
*** 1059,1066 ****
  	// associated subnet mask.
  	// e.g.: "192.0.2.5/24" for IPv4 and "2001:db8::5/64" for IPv6.
  	//
- 	// Must not contain more than 16 entries.
- 	//
  	// +optional
  	// +listType=atomic
  	IPs []string
--- 1057,1062 ----
```

- Add named constants for size limits (this number of entries of `IPs`, but also everything else checked by validation).

- Verify that `claim.status.devices.data` really is omitted when not set. It probably needs to be changed to `Data *runtime.RawExtension`.

/assign @LionelJouin
/sig node
/wg device-management

### Why is this needed?

Consistency.

**Comment by LionelJouin** [comment]
Thank you, PR created: https://github.com/kubernetes/kubernetes/pull/129158

**Comment by LionelJouin** [comment]
The third point is about using a pointer on the `Data` field, is it required? The validation checks if the length of `Data.Raw` is 0. Here is the discussion on slack during the initial PR:
https://kubernetes.slack.com/archives/CJUQN3E4T/p1730109344106609

**Comment by pohly** [comment]
I think a pointer is required. You can test by bringing up a cluster, creating a claim, adding a status *without* data and then look at the `kubectl get -o yaml` output: we don't want to see `Data`, but it's probably going to be there.

**Comment by LionelJouin** [comment]
isn't it a breaking change?

**Comment by pohly** [comment]
No, changing the Go representation of the serialized data is not a breaking change. We cannot distinguish between "not set" and "empty" because we serialized "not set" as "empty" in the stored protobuf in 1.32, but we don't need that distinction.

What matters is that non-empty `Data` gets encoded, decoded and stored as it was before.
